cli/lsp/language_server.rs

```diff
@@ -985,28 +985,28 @@ impl Inner {
 
     let mark = self.performance.mark("document_symbol", Some(&params));
     let asset_or_document = self.get_cached_asset_or_document(&specifier)?;
-    let line_index = asset_or_document.line_index();
-
-    let req = tsc::RequestMethod::GetNavigationTree(specifier);
-    let navigation_tree: tsc::NavigationTree = self
-      .ts_server
-      .request(self.snapshot()?, req)
-      .await
-      .map_err(|err| {
-        error!("Failed to request to tsserver {}", err);
-        LspError::invalid_request()
+    // we need to clone the line index _before_ the async process of grabbing
+    // the nav tree, as the line index could change
```

Comment:
I think there are some parts of the LSP that I don't understand. How does the line index change while getting the navigation tree?

Reply:
I think you are right... since we don't mutate the `Arc`... I need to re-think this... I did find one spot where we were getting the line index _after_ the call to tsc, which could be the issue, instead of holding onto the reference during it.

It is a real challenge to reproduce though, because it is a race condition, but I have seen some really strange behaviour on 1.16 that is consistent with the line_index being "wrong" that didn't exist before.

Reply:
Yeah, let's maybe tackle fixing #12776 and we can look more to try to get a reproduction of this issue.